### ISSUE TITLE
Changelog: commit hashes don't have a link with default template, closed issues are not referenced

### ISSUE THREAD

**Issue by gotson** [convco/convco#47]
**Describe the bug**
I am trying to generate the changelog for https://github.com/junrar/junrar but the generated changelog doesn't have links for commit hashes.

I noticed the [default template](https://github.com/convco/convco/blob/5712dae9aea8b2079ad0f6f1ef902da114ef249a/src/conventional/changelog/commit.hbs#L4) has a `{{~#if @root.linkReferences}}` but that doesn't seem to be working. Searching for `linkReferences` in the codebase yields no results.

**To Reproduce**
1. Clone https://github.com/junrar/junrar
2. Run `convco changelog -s -m 2`
3. The changelog contains:

```
# Changelog

## [v7.5.0](https://github.com/junrar/junrar/compare/v7.4.1...v7.5.0) (2022-03-20)


### Features

* parse extended time
 d5cc784

* use thread pool for getInputStream
 a3383b0


### Fixes

* crc errors on audio data decompression
 15f4afa

* NPE on audio data decompression
 952436b


### [v7.4.1](https://github.com/junrar/junrar/compare/v7.4.0...v7.4.1) (2022-01-27)


#### Fixes

* invalid subheader type would throw npe and make the extract loop
 7b16b3d
```

**Expected behavior**
I would expect something closer to:

```
# [7.5.0](https://github.com/junrar/junrar/compare/v7.4.1...v7.5.0) (2022-03-20)

### Features

* parse extended time ([d5cc784](https://github.com/junrar/junrar/commit/d5cc784c937f461be1e71a7a92b4018af8aef8c7))
* use thread pool for getInputStream ([a3383b0](https://github.com/junrar/junrar/commit/a3383b0dc4db5c5c29334abadd42688fa5ea583b))

### Bug Fixes

* crc errors on audio data decompression ([15f4afa](https://github.com/junrar/junrar/commit/15f4afa23eff69c2e0f3c906815777abf5ac267c))
* NPE on audio data decompression ([952436b](https://github.com/junrar/junrar/commit/952436b204614a747fe8a401d213196cd326d818))

## [7.4.1](https://github.com/junrar/junrar/compare/v7.4.0...v7.4.1) (2022-01-27)


### Bug Fixes

* invalid subheader type would throw npe and make the extract loop ([7b16b3d](https://github.com/junrar/junrar/commit/7b16b3d90b91445fd6af0adfed22c07413d4fab7)), closes [#73](https://github.com/junrar/junrar/issues/73)
```

Note the commit hashes have links, and the closes issues are also referenced.

**System (please complete the following information):**
 - OS: Mac
 - Version: 10.15.7

**Additional context**
`convco` installed by Home Brew, version 0.3.9

**Comment by gotson** [comment]
Note also that even without the link to the commit hash, the hash is placed on a new line, instead of being inside rounded brackets on the same line.

**Comment by gotson** [comment]
I tried overriding the default `commit.hbs`, by changing this:

```
*{{#if scope}} **{{scope}}:**{{/if}} {{#if subject}}{{#each subject}}{{this}}
{{/each}}{{else}}{{header}}{{/if}}
{{~#if hash}}
  {{~#if @root.linkReferences}} ([{{shortHash}}]({{commitUrlFormat}}))
  {{~else}} {{shortHash}}{{/if}}
{{~/if}}

{{#if references}}, closes{{#each references}} {{#if @root.linkReferences}}[{{#if this.owner}}{{this.owner}}/{{/if}}{{this.repository}}{{this.prefix}}{{this.issue}}]({{issueUrlFormat}}){{else}}{{#if this.owner}}{{this.owner}}/{{/if}}{{this.repository}}{{this.prefix}}{{this.issue}}{{/if}}{{/each}}{{/if}}
```

to this:

```
*{{#if scope}} **{{scope}}:**{{/if}} {{#if subject}}{{#each subject}}{{this}}
{{/each}}{{else}}{{header}}{{/if}}
{{~#if hash}}
  ([{{shortHash}}]({{commitUrlFormat}}))
{{~/if}}

{{#if references}}, closes{{#each references}} {{#if @root.linkReferences}}[{{#if this.owner}}{{this.owner}}/{{/if}}{{this.repository}}{{this.prefix}}{{this.issue}}]({{issueUrlFormat}}){{else}}{{#if this.owner}}{{this.owner}}/{{/if}}{{this.repository}}{{this.prefix}}{{this.issue}}{{/if}}{{/each}}{{/if}}
```

but i get the following error:

`Error rendering "commit" line 4, col 20: Variable "commitUrlFormat" not found in strict mode.`

**Comment by hdevalke** [comment]
Thanks for reporting.
I see 2 points you would like to fix.

1. Adding links to hashes by default.
2. commitUrlFormat should be available  at the place you tried to use it.

for 1. most web UIs make automatic urls for commits. Having no link makes the raw file more readable (e.g. when executing convco on the command line)

for 2. I will check if this is a bug in convco or a bug in your changes or a bug in handelbars-rs.

**Comment by hdevalke** [comment]
`commitUrlFormat` is available as `@root.commitUrlFormat`

links to hashes should be configurable with `linkReferences: true`, but that is not implemented.

I don't have much free time, but will check to implement it in the coming month, unless you want me to guide you to make a pull request.

**Comment by gotson** [comment]
> `commitUrlFormat` is available as `@root.commitUrlFormat`

In that case the default template is incorrect? I was just trying to force display the reference, since `linkReferences` didn't seem to work (and you confirmed that).

> for 1. most web UIs make automatic urls for commits. Having no link makes the raw file more readable (e.g. when executing convco on the command line)

Github doesn't seem to be, or maybe it's just me. I also see interest when taking the file out of context, in which case the implicit links may not work properly anymore. I think having it as a default would be a good default.

--- 

The "closes <issue numner>" also don't seem to work properly by default. Not sure how `references` is computed exactly. For instance in my example the commit in `7.4.1` closes issue 73, but that doesn't appear in the generated changelog.

**Comment by gotson** [comment]
> `commitUrlFormat` is available as `@root.commitUrlFormat`

I confirm that it works 👍🏻

**Comment by gotson** [comment]
Spoke too soon :(

`commitUrlFormat: "{{host}}/{{owner}}/{{repository}}/commit/{{hash}}"`

```
{{~#if hash}}
  ([{{shortHash}}]({{@root.commitUrlFormat}}))
{{~/if}}
```

this yields: 

```
* parse extended time
  ([d5cc784](https://github.com/junrar/junrar/commit/))
```

Looks like the `hash` replacement doesn't work as expected. I'm not very familiar with Mustache/Handlebars so not sure what's the problem.

**Comment by gotson** [comment]
> links to hashes should be configurable with `linkReferences: true`, but that is not implemented.
> 
> I don't have much free time, but will check to implement it in the coming month, unless you want me to guide you to make a pull request.

Would be happy to try. I've basic concepts of Rust, that would be a good occasion to start!

I also wanted to add unit tests for the changelog, to verify the parsing of issue references for the `closes` function, but i am not sure where to start.

**Comment by gotson** [comment]
> The "closes " also don't seem to work properly by default. Not sure how `references` is computed exactly. For instance in my example the commit in `7.4.1` closes issue 73, but that doesn't appear in the generated changelog.

I've been trying to find why this doesn't work, here are my findings:
- When parsing a footer like `refs #133`, it returns a `Footer{ key: "refs", value: 133 }` (as the [existing test](https://github.com/convco/convco/blob/712f455e76cee210c22389148a7d1793d784485a/src/conventional/commits.rs#L379-L382) shows).
- the changelog transformer is [running the regex](https://github.com/convco/convco/blob/1704b184b6cb62157c393ed3e0cdc2c941b87258/src/cmd/changelog.rs#L188) against `footer.value`, which by default won't work as the regex looks for `#`, but the `footer.value` doesn't contain the `#` anymore.

I would say that the `footer.value` should still contain the `#`. What do you think?

**Comment by hdevalke** [comment]
> I'm not very familiar with Mustache/Handlebars so not sure what's the problem.

me neither, I had issues to fix whitespace issues.

> I would say that the footer.value should still contain the #. What do you think?

The footer is separated by either `:<space>` or `<space>#` see also https://github.com/convco/convco/issues/48#issuecomment-1089247132

The `#` for the issue regex is meant to be definable by the user.

So `closes: #123` will work fine, but `closes #123` will not.

The idea of the reference regex is that a repo could have an external issue system (e.g. jira) having references like `JIRA-123` instead of `#123`. 

I would suggest to always use a footer with the `:<space>` separator

**Comment by gotson** [comment]
> I would suggest to always use a footer with the `:<space>` separator

but that's a bit of a bummer for existing repos with existing history.

Tools like conventional-changelog or semantic-release can parse those fine.

**Comment by gotson** [comment]
> The footer is separated by either `:<space>` or `<space>#`

I do agree, that's in the spec.

> The `#` for the issue regex is meant to be definable by the user.

I agree, but the defaults don't work in that case. There is `#` as default for the issue prefix, but the changelog will not pick them up, because the parser removed the `#`. Not very good on the UX.

> So `closes: #123` will work fine, but `closes #123` will not.

`closes #123` should work, it's part of the spec. I think the problem is that you only keep the key/value and trash the separator. You should keep the whole string in the `Footer` struct, and that is what the changelog should match against.

That way you could match `closes #`, `fixes #`, but not `#`.

> The idea of the reference regex is that a repo could have an external issue system (e.g. jira) having references like `JIRA-123` instead of `#123`.

Those are 2 different things.

If you use `#`, then you don't need the `:`. If you use `JIRA-123`, then you need the `:`.

`closes #123`
`closes: JIRA-123`

That's what the spec mentions:

> Each footer MUST consist of a word token, followed by either a `:<space>` or `<space>#`

**Comment by hdevalke** [comment]
I'll fix in this issue the `linkReferences` as the other issues described here are fixed in #48.

**Comment by hdevalke** [comment]
Thanks for reporting this. It was a bit harder to fix than first thought.

@gotson the example (`convco changelog -s -m 2`) now prints out:

```
# Changelog

## [v7.5.0](https://github.com/junrar/junrar/compare/v7.4.1...v7.5.0) (2022-03-20)

### Features

* parse extended time
  ([d5cc784](https://github.com/junrar/junrar/commit/d5cc784c937f461be1e71a7a92b4018af8aef8c7))
* use thread pool for getInputStream
  ([a3383b0](https://github.com/junrar/junrar/commit/a3383b0dc4db5c5c29334abadd42688fa5ea583b))

### Fixes

* crc errors on audio data decompression
  ([15f4afa](https://github.com/junrar/junrar/commit/15f4afa23eff69c2e0f3c906815777abf5ac267c))
* NPE on audio data decompression
  ([952436b](https://github.com/junrar/junrar/commit/952436b204614a747fe8a401d213196cd326d818))
```
